### PR TITLE
Track card reader battery property card reader events

### LINKED ISSUE #7651: Log card reader battery level when logging errors
**Describe the bug**
A clear and concise description of what the bug is. Please be as descriptive as possible; issues lacking detail, or for any other reason than to report a bug, may be closed without action.

It would be helpful to analyze the IPP support queries if we log the card reader battery level along with the error information. 

Stripe terminal SDK sometimes throws a [cryptic error](p91TBi-9sD-p2#comment-10903) when the card reader battery is low and this would be confusing to figure out where the problem lies.

**Expected behavior**
The card reader battery level would be logged along with the card reader errors

### PR DESCRIPTION
**Pull request by samiuelson**
Logging card reader's battery level tracking card reader events.

Closes: #7651

### Description
1. This PR adds the `CardReaderTrackingInfoKeeper::setCardReaderBatteryLevel` method for storing battery-level data in memory. 
2. Battery level is added as a property to events related to the card reader (tracked using `CardReaderTracker::track` method). The name of the event's property: `card_reader_battery_level`.
3. The battery level gets updated immediately after the reader is connected and then during the payment flow we subscribe to battery level changes and update it continuously using `CardReaderTrackingInfoKeeper`.

### Testing instructions
**1. Observe that events related to the card reader include the `card_reader_battery_level`.** 
Note: it won't be included if there are no battery-level data available. Usually, this shouldn't happen.
**2. Observe that the battery level is updated.** 
Eg. This can happen during long-running card payment flow if the reader is not connected to the cable. In that case, the battery level tracked with `*_card_reader_connection_success` event would differ from `*_card_present_collect_payment_success`. Note: the time from connecting to a reader to tapping a card must be long enough to observe battery level change.

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderTracker.kt [hunk 1]
```diff
@@ -91,6 +92,13 @@ class CardReaderTracker @Inject constructor(
         }
     }
 
+    private fun addCardReaderBatteryLevelProperty(properties: MutableMap<String, Any>) {
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderTracker.kt:95]: ❓ We could write tests to verify that the card reader tracker adds battery level property when the battery status is not null and doesn't add battery level property when the battery status is null in `CardReaderTrackerTest`. WDYT?

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderPaymentViewModelTest.kt [hunk 2]
```diff
@@ -2943,6 +2946,25 @@ class CardReaderPaymentViewModelTest : BaseUnitTest() {
 
     //endregion - Interac Refund tests
 
+    @Test
+    fun `when new battery status event is received, then tracking is updated with new battery level`() =
```
Comment (AnirudhBhat) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderPaymentViewModelTest.kt:2951]: ❓ We could write another test that verifies that the `cardReaderTrackingInfoKeeper.setCardReaderBatteryLevel` won't be called if the `CardReaderBatteryStatus` is anything but `StatusChanged`. WDYT?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderTrackingInfoProvider.kt [hunk 3]
```diff
@@ -41,13 +42,18 @@ class CardReaderTrackingInfoImpl @Inject constructor() : CardReaderTrackingInfoP
     override fun setCardReaderModel(cardReaderModel: String?) {
         trackingInfoInternal = trackingInfoInternal.copy(cardReaderModel = cardReaderModel)
     }
+
+    override fun setCardReaderBatteryLevel(batteryLevel: Float?) {
+        trackingInfoInternal = trackingInfoInternal.copy(cardReaderBatteryLevel = batteryLevel)
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderTrackingInfoProvider.kt:47]: 💡 Sorry, I forgot to mention this during our pair programming session. The battery level will be displayed in float and might not be readable when analyzing the events. For example: `"card_reader_battery_level":0.05`

We have an extension function that converts the battery level from the float into a readable string. It's in the `CardReaderDetailViewModel` class. The extension is named `getReadersBatteryLevel()`. We could probably use this.

We might need to extract the core battery conversion logic of float to int into a separate reusable method though since the above extension method appends the string "battery" for the battery level value. Example - `5% battery`

WDYT?

**Comment by AnirudhBhat** [comment]
Thanks for working on this @samiuelson The functionality works as expected 🎉 

I've left a few minor comments. Again, Great job on this!

**Comment by samiuelson** [comment]
> Thanks for working on this @samiuelson The functionality works as expected 🎉
> 
> I've left a few minor comments. Again, Great job on this!

Thanks for the review @AnirudhBhat! I've added suggested changes 👍🏼

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderUtils.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package com.woocommerce.android.ui.payments.cardreader
+
+import com.woocommerce.android.R
+import com.woocommerce.android.cardreader.connection.CardReader
+import com.woocommerce.android.model.UiString
+import kotlin.math.roundToInt
+
+private const val PERCENT_100 = 100
+private fun Float.toPercent(): Int = (this * PERCENT_100).roundToInt()
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderUtils.kt:9]: 👍

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderTrackerTest.kt [hunk 5]
```diff
@@ -101,6 +103,17 @@ class CardReaderTrackerTest : BaseUnitTest() {
             )
         }
 
+    @Test
+    fun `given battery level is null, when event is tracked, then it shouldn't contain battery level property`() =
```
Comment (AnirudhBhat) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderTrackerTest.kt:110]: ❓ We could write another test that verifies that the battery level property is included when the battery level is not null. WDYT?

```
@Test
    fun `given battery level is not null, when event is tracked, then it should contain battery level property`() =
        testBlocking {
            val cardReaderTrackingInfo: TrackingInfo = mock {
                on { cardReaderBatteryLevel } doReturn 0.5F
            }
            whenever(cardReaderTrackingInfoProvider.trackingInfo).thenReturn(cardReaderTrackingInfo)
            val props = mutableMapOf<String, Any>()

            cardReaderTracker.track(CARD_PRESENT_COLLECT_PAYMENT_SUCCESS, props)

            assertTrue(props.containsKey("card_reader_battery_level"))
        }
```

**Review by AnirudhBhat — APPROVED** [review]
Thanks for the changes @samiuelson. I've left one minor comment. LGTM!
